subsys/net/l2/ethernet/arp.c

```diff
@@ -259,6 +259,12 @@ static inline struct net_pkt *arp_prepare(struct net_if *iface,
 						sizeof(struct net_arp_hdr),
```

Comment:
The commit message bullet point 2. is hard to understand as there are typos and complex sentences.
I tried to fix the first part of the text
```
2. in ethernet_send(), if ARP packets send failed, original code tried to remove the
entry from ARP pending queue, but the entry wasn't removed from
the arp_pending_entries until the ARP request was timeout.
Within this timeout period, this ARP pending entry can't be found due to
the address mismatch.
```
Does the above look ok?

But I could not understand what the rest of the text is saying. Could you try to elaborate the text more?
```
But if add the entry remove and append action here, it will induce more
risk, as it needs to add arp_mutex protection, but net_arp_update()
has arp_mutex already and it will also call the ethernet_send. So the
better way is just remove the net_arp_clear_pending() here and let
the ARP request timeout do the entry cleanup and remove at same time.
```

Reply:
Your fix of the first part of the text looks good.
The rest of the text:
If add ‘entry removed from arp_pending_entries and append to arp_free_entries' here, it should also add arp_mutex protection like other functions in arp.c. But add arp_mutex protection will have side effect. For example, in the net_arp_update(), it will take the arp_mutex first, then may call the ethernet_send(), so add other arp_mutex protection in ethernet_send() is not a good choice and may lead to dead lock.
The better way is just removing the net_arp_clear_pending() in ethernet_send(), and let the ARP request timeout to do the entry cleanup, removed and append to arp_free_entries at the same time.